File: src/polymesh/a2a/poller.py
"""Status polling with the normative A2A backoff schedule (§A.10).

``delay = max(0, base + base * Uniform(-0.20, +0.20))`` where
``base = min(500 * 2**n, 15000)``.  Polling stops on a terminal state, the mesh
deadline, or cancellation; the wait itself is interruptible.
"""

from __future__ import annotations

import asyncio
import random
import time
from collections.abc import Awaitable, Callable, Mapping
from typing import Any, Protocol

from .errors import A2AError
from .task_translator import MonotonicStateGate, is_terminal_a2a_state, translate_task_event
from .types import TranslatedTaskEvent

POLL_BASE_MS = 500
POLL_MAX_MS = 15000
POLL_JITTER_RATIO = 0.20


class _Rng(Protocol):
    def uniform(self, a: float, b: float) -> float: ...


def compute_poll_delay(n: int, rng: _Rng | None = None) -> float:
    """Poll delay in milliseconds for zero-based attempt index ``n`` (§A.10.2)."""

    if not isinstance(n, int) or isinstance(n, bool) or n < 0:
        raise ValueError("poll attempt index must be a non-negative integer")
    # 2 ** n on a large n is cheap in Python but pointless past the cap.
    base = POLL_MAX_MS if n >= 32 else min(POLL_BASE_MS * (2**n), POLL_MAX_MS)
    source = rng if rng is not None else random
    jitter = base * source.uniform(-POLL_JITTER_RATIO, POLL_JITTER_RATIO)
    return max(0.0, base + jitter)


def _now_ms() -> int:
    return int(time.time() * 1000)


async def _interruptible_sleep(
    delay_ms: float,
    *,
    sleep: Callable[[float], Awaitable[None]],
    cancel_event: asyncio.Event | None,
) -> bool:
    """Sleep ``delay_ms``; return True when cancellation interrupted the wait."""

    if cancel_event is not None and cancel_event.is_set():
        return True
    if delay_ms <= 0:
        return cancel_event is not None and cancel_event.is_set()
    if cancel_event is None:
        await sleep(delay_ms / 1000.0)
        return False
    waiter = asyncio.ensure_future(cancel_event.wait())
    sleeper = asyncio.ensure_future(sleep(delay_ms / 1000.0))
    try:
        done, pending = await asyncio.wait({waiter, sleeper}, return_when=asyncio.FIRST_COMPLETED)
    finally:
        for task in (waiter, sleeper):
            if not task.done():
                task.cancel()
    await asyncio.gather(*pending, return_exceptions=True)
    return waiter in done


async def poll_until_terminal(
    *,
    fetch: Callable[[], Awaitable[Mapping[str, Any]]],
    task_id: str,
    deadline_ms: int | None = None,
    cancel_event: asyncio.Event | None = None,
    on_event: Callable[[TranslatedTaskEvent], None] | None = None,
    gate: MonotonicStateGate | None = None,
    rng: _Rng | None = None,
    sleep: Callable[[float], Awaitable[None]] | None = None,
    now_ms: Callable[[], int] | None = None,
    max_attempts: int | None = None,
) -> tuple[Mapping[str, Any], int]:
    """Poll ``fetch`` until the remote task is terminal (§A.10.3).

    Returns the terminal task snapshot and the number of polls performed.
    """

    clock = now_ms or _now_ms
    sleeper = sleep or asyncio.sleep
    state_gate = gate if gate is not None else MonotonicStateGate()
    attempt = 0
    poll_count = 0

    while True:
        if cancel_event is not None and cancel_event.is_set():
            raise A2AError("TASK_CANCELLED", "outbound A2A task cancelled by caller", task_id=task_id)
        if deadline_ms is not None and clock() >= deadline_ms:
            raise A2AError("DEADLINE", "outbound A2A task exceeded its mesh deadline", task_id=task_id)

        snapshot = await fetch()
        poll_count += 1
        state = snapshot.get("status", {}).get("state") if isinstance(snapshot, Mapping) else None
        if state_gate.accept(str(state)):
            event = translate_task_event(snapshot, task_id=task_id)
            if on_event is not None:
                on_event(event)
        if is_terminal_a2a_state(state):
            return snapshot, poll_count

        if max_attempts is not None and attempt + 1 >= max_attempts:
            raise A2AError("DEADLINE", "outbound A2A task exceeded its poll budget", task_id=task_id)

        delay_ms = compute_poll_delay(attempt, rng)
        if deadline_ms is not None:
            remaining = deadline_ms - clock()
            if remaining <= 0:
                raise A2AError("DEADLINE", "outbound A2A task exceeded its mesh deadline", task_id=task_id)
            delay_ms = min(delay_ms, float(remaining))
        if await _interruptible_sleep(delay_ms, sleep=sleeper, cancel_event=cancel_event):
            raise A2AError("TASK_CANCELLED", "outbound A2A task cancelled by caller", task_id=task_id)
        attempt += 1


__all__ = [
    "POLL_BASE_MS",
    "POLL_JITTER_RATIO",
    "POLL_MAX_MS",
    "compute_poll_delay",
    "poll_until_terminal",
]

File: packages/a2a/src/poller.ts
/**
 * Outbound status polling with bounded exponential backoff
 * (PM-V6-SPEC §A.10.2–§A.10.3, §A.10.6).
 */
import { A2ADialectError } from "./errors.js";
import { MonotonicStateTracker, isTerminalA2AState, translateTaskEvent } from "./task-translator.js";
import type { TranslatedTaskEvent } from "./task-translator.js";
import type { A2ATask } from "./types.js";

/** Initial delay unit before jitter (§A.10.2). */
export const POLL_BASE_MS = 500;
/** Backoff cap before jitter (§A.10.2). */
export const POLL_MAX_MS = 15_000;
/** Jitter fraction; delay is base ± 20% (§A.10.2). */
export const POLL_JITTER_FRACTION = 0.2;

/**
 * `base = min(500 * 2^n, 15000); jitter = base * Uniform(-0.20, +0.20);
 * delay = max(0, base + jitter)` (§A.10.2).
 *
 * `random` MUST return a value in `[0, 1)`; injecting it makes the schedule
 * deterministic for conformance tests.
 */
export function computePollDelay(
  n: number,
  random: () => number = Math.random,
  maxMs: number = POLL_MAX_MS,
): number {
  const attempt = Math.max(0, Math.floor(n));
  const uncapped = POLL_BASE_MS * Math.pow(2, Math.min(attempt, 52));
  const base = Math.min(uncapped, maxMs);
  const jitter = base * ((random() * 2 - 1) * POLL_JITTER_FRACTION);
  return Math.max(0, base + jitter);
}

/** Capped base for attempt `n`, ignoring jitter (§A.10.2 table). */
export function pollBaseDelay(n: number, maxMs: number = POLL_MAX_MS): number {
  const attempt = Math.max(0, Math.floor(n));
  return Math.min(POLL_BASE_MS * Math.pow(2, Math.min(attempt, 52)), maxMs);
}

export type SleepFn = (ms: number, signal?: AbortSignal) => Promise<void>;

/** Interruptible sleep; cancel MUST interrupt the wait early (§A.10.2). */
export const defaultSleep: SleepFn = (ms, signal) =>
  new Promise<void>((resolve, reject) => {
    if (signal?.aborted) {
      reject(pollCancelledError());
      return;
    }
    const timer = setTimeout(() => {
      cleanup();
      resolve();
    }, ms);
    const onAbort = (): void => {
      clearTimeout(timer);
      cleanup();
      reject(pollCancelledError());
    };
    function cleanup(): void {
      signal?.removeEventListener("abort", onAbort);
    }
    signal?.addEventListener("abort", onAbort, { once: true });
  });

export function pollCancelledError(): A2ADialectError {
  return new A2ADialectError("CANCELLED", "Outbound A2A poll cancelled by caller");
}

export function pollDeadlineError(): A2ADialectError {
  return new A2ADialectError("DEADLINE", "Outbound A2A task exceeded mesh deadline");
}

export interface PollUntilTerminalOptions {
  /** Issues one `tasks/get` against the remote. */
  getTask: (signal?: AbortSignal) => Promise<A2ATask>;
  /** Epoch-ms deadline; max poll duration equals the mesh task deadline (§A.10.2). */
  deadlineMs?: number;
  signal?: AbortSignal;
  now?: () => number;
  sleep?: SleepFn;
  random?: () => number;
  /** Backoff cap override (`poll_max_ms`, §E.13.1). */
  maxDelayMs?: number;
  /** Invoked for every accepted (non-stale) status transition. */
  onEvent?: (event: TranslatedTaskEvent, task: A2ATask) => void;
  /** Pre-seeded tracker so send-time ACCEPTED participates in monotonicity. */
  tracker?: MonotonicStateTracker;
}

export interface PollOutcome {
  task: A2ATask;
  event: TranslatedTaskEvent;
  poll_count: number;
  tracker: MonotonicStateTracker;
}

/**
 * `PollUntilTerminal` (§A.10.3). Stops on the first terminal A2A state,
 * throws `DEADLINE` at the deadline and `CANCELLED` on abort.
 */
export async function pollUntilTerminal(options: PollUntilTerminalOptions): Promise<PollOutcome> {
  const now = options.now ?? Date.now;
  const sleep = options.sleep ?? defaultSleep;
  const random = options.random ?? Math.random;
  const tracker = options.tracker ?? new MonotonicStateTracker("ACCEPTED");
  const signal = options.signal;

  let n = 0;
  let pollCount = 0;

  for (;;) {
    if (signal?.aborted) throw pollCancelledError();
    if (options.deadlineMs !== undefined && now() >= options.deadlineMs) throw pollDeadlineError();

    const task = await options.getTask(signal);
    pollCount += 1;
    const event = translateTaskEvent(task);

    // §A.10.6: never regress; discard stale updates after terminal.
    const applied = tracker.apply(event.state);
    if (applied) options.onEvent?.(event, task);

    if (isTerminalA2AState(task.status.state) && applied) {
      return { task, event, poll_count: pollCount, tracker };
    }
    if (tracker.terminal) {
      return { task, event, poll_count: pollCount, tracker };
    }

    let delay = computePollDelay(n, random, options.maxDelayMs ?? POLL_MAX_MS);
    if (options.deadlineMs !== undefined) {
      const remaining = options.deadlineMs - now();
      if (remaining <= 0) throw pollDeadlineError();
      delay = Math.min(delay, remaining);
    }
    await sleep(delay, signal);
    n += 1;
  }
}
